File: fiforpc/fiforpc_client.py
#!/usr/bin/env python3

import fileinput
import os

SERVER_TO_CLIENT_PATH = "/tmp/server_to_client.fifo"
CLIENT_TO_SERVER_PATH = "/tmp/client_to_server.fifo"

def perform_query(s):
    with open(CLIENT_TO_SERVER_PATH, "w") as c2s:
        print(s, file=c2s)

    out = ""
    with open(SERVER_TO_CLIENT_PATH, "r") as s2c:
        out = s2c.readline()
    return out

def main():
    while True:
        line = ""
        try: line = input("> ")
        except: break
        line = line.strip()
        print("got this line: ", line)

        returned = perform_query(line)
        print("returned from server: ", returned)

if __name__ == "__main__": main()

File: fiforpc/fiforpc_client.cc
#include <iostream>
#include <fstream>
#include <string>

using std::string;

const string SERVER_TO_CLIENT_PATH = "/tmp/server_to_client.fifo";
const string CLIENT_TO_SERVER_PATH = "/tmp/client_to_server.fifo";

string perform_query(const string& s) {
  // Write query to server's input fifo.
  std::ofstream c2s;
  c2s.open(CLIENT_TO_SERVER_PATH.c_str());
  c2s << s;
  c2s.close();

  // Read response from server's output fifo.
  string out;
  std::ifstream s2c(SERVER_TO_CLIENT_PATH.c_str());
  std::getline(s2c, out);
  s2c.close();
  return out;
}

int main(int argc, char *argv[]) {
  string line;

  while(true) {
    bool gotline = std::getline(std::cin, line);
    if (!gotline) break;

    std::cout << "got this line: " << line << std::endl;
    string returned = perform_query(line);
    std::cout << "server returned: " << returned << std::endl;
  }
}
